### PR TITLE
Fixes #9236: Log levels should be incremental

### PR DESCRIPTION
**Pull request by amousset**
https://www.rudder-project.org/redmine/issues/9236

Replacing previous PR: https://github.com/Normation/ncf/pull/469

### CODE REVIEW

**Comment by amousset** [comment]
Commit modified

File: tree/10_ncf_internals/configuration.cf [hunk 1]
```diff
@@ -79,6 +79,18 @@ bundle agent configuration
       "pass2" expression => "pass1";
       "pass1" expression => "any";
 
+      # Make log level incremental
+      "debug"        expression => "trace",
+                          scope => "namespace";
+      # CFEngine's stdlib verbose mode
+      "verbose_mode" expression => "debug",
+                          scope => "namespace";
+      "info"         expression => "debug",
```
Comment (peckpeck) [inline tree/10_ncf_internals/configuration.cf]: shouldn' this be verbose not debug ?

**Comment by peckpeck** [comment]
everything seems reverted, debug implies verbose etc ...

**Comment by peckpeck** [comment]
Did you see this with jon ?

File: tree/10_ncf_internals/configuration.cf (hunk 1 above)
Comment (amousset) [inline tree/10_ncf_internals/configuration.cf]: the verbose class does not exist

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Changed some reports to make things consistent: info level is only displayed when the info class is defined. This is closer to CFEngine behavior.

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Also added some doc (in the quickstart as we do not have anything better for now).

File: tree/10_ncf_internals/configuration.cf [hunk 2]
```diff
@@ -79,6 +79,18 @@ bundle agent configuration
       "pass2" expression => "pass1";
       "pass1" expression => "any";
 
+      # Make log level incremental
+      "debug"        expression => "trace",
```
Comment (peckpeck) [inline tree/10_ncf_internals/configuration.cf]: I still see trace implies debug

File: tree/10_ncf_internals/configuration.cf [hunk 3]
```diff
@@ -79,6 +79,18 @@ bundle agent configuration
...
+                          scope => "namespace";
+      # CFEngine's stdlib verbose mode
+      "verbose_mode" expression => "debug",
```
Comment (peckpeck) [inline tree/10_ncf_internals/configuration.cf]: debug implies verbose

File: tree/10_ncf_internals/configuration.cf (hunk 1 above)
Comment (peckpeck) [inline tree/10_ncf_internals/configuration.cf]: debug implies info

Comment (amousset) [inline tree/10_ncf_internals/configuration.cf]: What do you think should it be instead?

Comment (ncharles) [inline tree/10_ncf_internals/configuration.cf]: verbose_mode is defined by running agent with -v ; so it should imply info

Comment (amousset) [inline tree/10_ncf_internals/configuration.cf]: Interesting!

Comment (amousset) [inline tree/10_ncf_internals/configuration.cf]: In ncf, we have:
* *info*
* *debug*
* *trace*

CFEngine has:
* *inform_mode*
* *verbose_mode*
* *debug_mode*

We should at least define ncf classes when CFEngine classes are defined.

Comment (peckpeck) [inline tree/10_ncf_internals/configuration.cf]: you're right i reversed the implication in my head

So
- we say that rudder loglevel decides cfengine stdlib loglevel
- we say that upper loglevel implies lower loglevel

Where does the trace loglevel come from ?

Comment (peckpeck) [inline tree/10_ncf_internals/configuration.cf]: So we should set debug_mode too here no ?

Comment (peckpeck) [inline tree/10_ncf_internals/configuration.cf]: I've just seen your upper comment.
Don't we need to split debug and trace too in rudder agent ?

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

File: tree/10_ncf_internals/configuration.cf [hunk 4]
```diff
@@ -79,6 +79,19 @@ bundle agent configuration
       "pass2" expression => "pass1";
       "pass1" expression => "any";
 
+      "info"         expression => "inform_mode",
```
Comment (amousset) [inline tree/10_ncf_internals/configuration.cf]: @peckpeck @ncharles Now:

* Options defined by CFEngine will define the matching ncf verbosity class (but verbose in CFEngine = debug in ncf, and debug in CFEngine = trace in ncf)
* ncf classes are incrementally defined

Comment (peckpeck) [inline tree/10_ncf_internals/configuration.cf]: Thanks, this is much easier to understand

Comment (peckpeck) [inline tree/10_ncf_internals/configuration.cf]: in this case it means our class definition in rudder agent is redundant ?

Comment (amousset) [inline tree/10_ncf_internals/configuration.cf]: Indeed, I did not know this class was set by CFEngine itself.

Comment (amousset) [inline tree/10_ncf_internals/configuration.cf]: We might want to keep it for now though, for compatibility with older ncf.

**Comment by amousset** [comment]
Commit modified

File: tree/10_ncf_internals/configuration.cf (hunk 4 above)
Comment (peckpeck) [inline tree/10_ncf_internals/configuration.cf]: you're right

**Pull request merged**